### DISCUSSION TITLE
How do people handle project where exact steps are not known in advance?

### DISCUSSION THREAD

**Discussion by mooskagh** [GothenburgBitFactory/taskwarrior#2804] [Q&A]
That's something that I've just posted in https://github.com/GothenburgBitFactory/taskwarrior/issues/2193#issuecomment-1133846839, but just realized that the discussion board is a better place for it.

When I work on tasks, it happens to me more often than not, that a particular task has to be split, yet almost all task trackers that I tried, don't support that easily (the only exception that I remember is AbstractSpoon TODO list, it's great, but it's unfortunately Windows only).

So typically when there is a need for a new project, I add some task like "Implement feature XXX in YYY". When I add it, I usually don't have enough desire to think (and knowledge) to tell what exactly to do.

Then when time comes, I start to add subtasks (usually 1-3 ahead, not the complete plan), something like "Research how it's done in similar projects". Then, when I decide to work on that task, I often split it into subtasks further, and so on, until tasks are doable within a day. Also more often than not, it happens, that you don't know in advance whether the task will have to be split. E.g. initially you think that you just add a library to your code and call a function, it will be a 10 minutes task, but then you end up fighting with compilation of this library for 2 days, fixing errors there and sending GitHub PRs upstream. In that case I split it into subtasks further. It's not rare for me to have "depth 5 subtask".

TW projects don't really help with that, as there's no concept of "splitting" and you kind of supposed to know most of the steps in advance. I've written a sophisticated wrapper around TW to handle that, but I wonder what other people do.

So, I wonder how do people approach that problem of having only vague tasks in the beginning and clarifying the next steps as the progress goes?

Thanks!

**Comment by angelus2014** [comment]
Usually, I create a task 'Set up project for XYZ'; where the project will house the various tasks which I couldn't think of at first. I _do_ miss a function like 'convert to project' that applications like OF and Things3 have. I also use the 'depends' option, but it's very inflexible as it doesn't change when the task numbers change; certain 'types' of projects (parallel, sequential, etc.) would be greatly appreciated.

**Reply by djmitche** [reply]
> doesn't change when the task numbers change

Can you explain what you mean?  If task X depends on task Y, how does that relate to the task numbers for X and Y?

**Reply by angelus2014** [reply]
I use project templates by duplicating. If there is a dependency, the new tasks will have new task numbers, but they're pointing to the old dependency task numbers.

With inflexible I mean: I have a project (a sequential one) to which I have (already a bit of a pain in TW) added dependencies. But as it often happens, there is another task that comes up later, but in the dependency 'tree' is somewhere in the middle, or at 3rd place, or where ever. Or some external factor makes it necessary to change the order of certain tasks. Stuff that is quite easy to do in a task/project manager that supports proper sequential projects, but not so in TW I'm sorry to say. For me personally, the dependency setup (or rather lack of sequential projects) is a big issue in TW only surpassed by the way TW handles recurring tasks.

**Comment by jbalcorn** [comment]
I do the same - start with "Set up Project X".  I also create a "Project X
Complete" task that depends on the first.  Then as I determine what
steps/tasks are necessary, I put them in, using "depend" as needed, and
"blocks:" on the final task using
https://gist.github.com/wbsch/a2f7264c6302918dfb30

If the "Project X Complete" shows up without it being done, I know I have
tasks to add to taskwarrior.


-- 
Justin B. Alcorn
The views expressed are not necessarily my own, much less anyone else's
PGP Fingerprint CCEB F776 C3FD 1050 C8DB  532E B8B9 BED7 7764 406C

**Comment by jbalcorn** [comment]
When I use depends the linkage is reliable as I think it links on UUID.
Are you getting some other kind of behavior?
-- 
Justin B. Alcorn
The views expressed are not necessarily my own, much less anyone else's
PGP Fingerprint CCEB F776 C3FD 1050 C8DB  532E B8B9 BED7 7764 406C

**Comment by djmitche** [comment]
One general answer to this question is that interfaces from TW to other task/project-tracking systems can be helpful.  For example, BugWarrior supports converting GH issues to TW tasks.  So I can do project-management kinds of things in GitHub (using its PM tools), and then have BW sync the issues assigned to me into TaskWarrior.  This lets TW focus on its strength -- "what's next" -- while leveraging other tools (GitHub is just one example) for their strengths around project management, collaboration, and so on.

And, I might have a task in TW with a title like "Set up Jira board for new Spinaga initiative" to track actually _doing_ the initial PM work, of course.

**Comment by smemsh** [comment]
it would be nice if tasks could have familial relations and forking as native concepts.  Instead `project` is really just an extra string label.  It is not any kind of lineage relationship that has referential integrity maintained by the system. Having this and a few commands around like fork/merge/splice would allow "real" subtasks
